### PR TITLE
Improve async/await usage, and sort init calls in `index.js`

### LINKED ISSUE #17448: Wrong async/await/preventDefault usage
This is an issue for discussion of frontend refactoring. @lafriks @silverwind 

The code https://github.com/go-gitea/gitea/blob/main/web_src/js/features/repo-legacy.js#L350 works like this:

```
$(document).on('click', '.edit-content', async function (event) {
    await createDropzone();
    event.preventDefault();
});

// equal to 
$(document).on('click', '.edit-content', function (event) {
    return Promise.resolve(createDropzone).then(() => {
        event.preventDefault();
    });
});
```

The usage of async/await is incorrect. The `preventDefault` won't have effect because the event has been processed synchronously before (when the listener returned).

So, that's why I insist to clean up most inconsistent await/async in the code base, only use them when it is necessary.

* https://github.com/go-gitea/gitea/pull/17386

We should do our best to make everything work as expected.

image

### PR DESCRIPTION
**Pull request by wxiaoguang**
The usage of async/await in some cases are incorrect. This PR makes the async/await more clear and sort the `initXxx` calls in `index.js` as planned in https://github.com/go-gitea/gitea/pull/17315 .

This PR only moves code, renames some functions, and processes async/await. No change to logic (hopefully)

==

Background for async/await in JavaScript:

```
async function foo() {
    func1();
    await bar();
    func2();
}
// equals to
function foo() {
    func1();
    return Promise.resolve(bar).then(() => {
         func2();
    });
}
```

So, many async/await in old code have no effect, and won't do `parallel init of async loaded features`, these functions are still executed one by one because JavaScript engines in browsers are single-threaded. Removing these async keywords makes the init logic more simple.

### CODE REVIEW

**Comment by silverwind** [comment]
> these functions are still executed one by one because JavaScript engines in browsers are single-threaded. Removing these async keywords makes the init logic more simple.

These functions do run concurrently and there's certainly a speed benefit of long-running operations like chunk loading. 

I think not `await`ing an async function is an anti-pattern because it breaks the linearity of code execution. Async function immediately return a promise and then continue to execute their code, which in itself can be a source of bugs.

File: web_src/js/features/common-global.js [hunk 1]
```diff
@@ -135,11 +135,11 @@ export function initGlobalCommon() {
   });
 }
 
-export async function initGlobalDropzone() {
+export function initGlobalDropzone() {
```
Comment (silverwind) [inline web_src/js/features/common-global.js:138]: I'm strongly against this, dropzone will not be initialized after `initGlobalDropzone` has completed and there function consumer won't be able to know when it has.

File: web_src/js/features/dropzone.js [hunk 2]
```diff
@@ -1,9 +1,6 @@
 export default async function createDropzone(el, opts) {
-  const [{Dropzone}] = await Promise.all([
-    import(/* webpackChunkName: "dropzone" */'dropzone'),
-    import(/* webpackChunkName: "dropzone" */'dropzone/dist/dropzone.css'),
-  ]);
-
+  import(/* webpackChunkName: "dropzone" */'dropzone/dist/dropzone.css');
+  const {Dropzone} = await import(/* webpackChunkName: "dropzone" */'dropzone');
```
Comment (silverwind) [inline web_src/js/features/dropzone.js]: This can introduce a bug that the function may return before `dropzone.css` has been loaded. I'm strongly against this.

File: web_src/js/features/notification.js [hunk 3]
```diff
@@ -66,7 +66,7 @@ export async function initNotificationCount() {
         return;
       }
       if (event.data.type === 'notification-count') {
-        receiveUpdateCount(event.data);
+        const _ = receiveUpdateCount(event.data);
```
Comment (silverwind) [inline web_src/js/features/notification.js]: Pointless assignment imho.

File: web_src/js/features/repo-editor.js [hunk 4]
```diff
@@ -75,7 +75,7 @@ function getCursorPosition($e) {
   return pos;
 }
 
-export async function initRepoEditor() {
+export function initRepoEditor() {
```
Comment (silverwind) [inline web_src/js/features/repo-editor.js:78]: Again, function will return before editor is initialized, I don't like it.

**Comment by wxiaoguang** [comment]
> > these functions are still executed one by one because JavaScript engines in browsers are single-threaded. Removing these async keywords makes the init logic more simple.
> 
> These functions do run concurrently and there's certainly a speed benefit of long-running operations like chunk loading.

How do they run concurrently and how it improves speed? I didn't see any difference with the old code. There is no long-running operations during in `index.js` and the `init` functions. Every `init` function can do its private long-running operation inside, but not in a global `$(document).ready()`.

If there were some benefits, we need real examples and evidences.

> I think not `await`ing an async function is an anti-pattern because it breaks the linearity of code execution. Async function immediately return a promise and then continue to execute their code, which in itself can be a source of bugs.

What it breaks, I do not find the breakings you worried. When I remove async I also remove unnecessary await. In fact, in old code, there are async functions called without await. I just made things more clear.

File: web_src/js/features/common-global.js (hunk 1 above)
Comment (wxiaoguang) [reply web_src/js/features/common-global.js:138]: What are these function consumers? If there is code depending on `initGlobalDropzone` but appears in other modules, we should make them work together, we can not make coupling code everywhere.

If there is no other function consumer code, then this change is OK.

How do you think?

File: web_src/js/features/dropzone.js (hunk 2 above)
Comment (wxiaoguang) [reply web_src/js/features/dropzone.js]: In my mind, the CSS loading is not necessary this change can improve speed slightly. 

If you worry about the CSS loading, I will revert it.

File: web_src/js/features/repo-editor.js (hunk 4 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-editor.js:78]: I need the real reasons why it doesn't work. There should be no dependency between modules.

Most modules (features) should be able to be initialized at any order.

If there was a dependency, we should make it clear now. I can not bear the noodle-like logic.

File: web_src/js/features/notification.js (hunk 3 above)
Comment (wxiaoguang) [reply web_src/js/features/notification.js]: It is not pointless. `receiveUpdateCount` is async, but we do not do async things here. 

The `const _` just indicates that we do not care about the Promise returned by `receiveUpdateCount`, and makes IDE lint happy.

Comment (wxiaoguang) [reply web_src/js/features/notification.js]: Maybe we can use `const _promise` then it's more clear.

Comment (delvh) [reply web_src/js/features/notification.js]: Did you just say … the linter is happier when there are unused variables lying around???
That sounds like a misconfigured linter to me…

File: web_src/js/features/common-global.js (hunk 1 above)
Comment (delvh) [reply web_src/js/features/common-global.js:138]: Wait, @silverwind, are you arguing that we cannot make `initGlobalDropzone` non-async because the completion listener in its own method won't be triggered in its own function?
That seems… counter-intuitive and confusing?
If both methods are called synchronously, then you don't have the problem that you need a completion listener.

File: web_src/js/features/common-global.js [hunk 5]
```diff
@@ -135,11 +135,11 @@ export function initGlobalCommon() {
...
   // Dropzone
   for (const el of document.querySelectorAll('.dropzone')) {
     const $dropzone = $(el);
-    await createDropzone(el, {
+    const _promise = createDropzone(el, {
```
Comment (delvh) [inline web_src/js/features/common-global.js:142]: I am against using unused variables.
I don't know why a linter would complain about unused returns instead of unused variables…

File: web_src/js/features/comp/SearchUserBox.js [hunk 6]
```diff
@@ -2,7 +2,7 @@ import {htmlEscape} from 'escape-goat';
 
 const {appSubUrl} = window.config;
 
-export function initSearchUserBox() {
+export function initCompSearchUserBox() {
```
Comment (delvh) [inline web_src/js/features/comp/SearchUserBox.js:5]: What's the meaning of the `Comp`?
And why was this changed?

File: web_src/js/features/notification.js [hunk 7]
```diff
@@ -66,7 +66,7 @@ export async function initNotificationCount() {
         return;
       }
       if (event.data.type === 'notification-count') {
-        receiveUpdateCount(event.data);
+        const _promise = receiveUpdateCount(event.data);
```
Comment (delvh) [inline web_src/js/features/notification.js:69]: I think my opinion on that is already clear.

File: web_src/js/features/repo-commit.js [hunk 8]
```diff
@@ -1,6 +1,47 @@
+const {csrfToken} = window.config;
+
 export function initRepoCommitButton() {
   $('.commit-button').on('click', function (e) {
     e.preventDefault();
     $(this).parent().find('.commit-body').toggle();
   });
 }
+
+export function initRepoCommitLastCommitLoader() {
```
Comment (delvh) [inline web_src/js/features/repo-commit.js:10]: What about `initLastRepoCommitLoader` or the original name `initLastCommitLoader`?

File: web_src/js/features/repo-diff.js [hunk 9]
```diff
@@ -79,3 +79,28 @@ export function initRepoDiffConversationNav() {
     window.location.href = `#${anchor}`;
   });
 }
+
+export function initRepoDiffShowMore() {
```
Comment (delvh) [inline web_src/js/features/repo-diff.js:83]: What about `initDiffShowMore`, as it was previously?

File: web_src/js/features/repo-editor.js [hunk 10]
```diff
@@ -24,7 +24,7 @@ function initEditPreviewTab($form) {
         _csrf: csrfToken,
         mode,
         context,
-        text: $form.find(`.tab[data-tab="${$tabMenu.data('write')}"] textarea`).val()
+        text: $form.find(`.tab[data-tab="${$tabMenu.data('write')}"] textarea`).val(),
```
Comment (delvh) [inline web_src/js/features/repo-editor.js:27]: Please don't tell me the linter complained again…

File: web_src/js/features/repo-issue-content.js [hunk 11]
```diff
@@ -104,7 +104,7 @@ function showContentHistoryMenu(issueBaseUrl, $item, commentId) {
   });
 }
 
-export function initIssueContentHistory() {
+export function initRepoIssueContentHistory() {
```
Comment (delvh) [inline web_src/js/features/repo-issue-content.js:107]: I am absolutely not a fan of adding `Repo` to all these methods. Same as the `Comp` seen earlier. Simply why?
I'd say it makes the code less readable instead of more readable.

File: web_src/js/features/repo-projects.js [hunk 12]
```diff
@@ -1,6 +1,6 @@
 const {csrfToken} = window.config;
 
-export default async function initProject() {
+export default async function initRepoProject() {
```
Comment (delvh) [inline web_src/js/features/repo-projects.js]: Which functions are async, and which are not?
It seems rather random which are, and which aren't.
That is indeed unmaintainable chaos.
Could we please remove the `async` everywhere once, and then only add it for those that are actually needed?
I would even recommend a separate package for async functions, as they are handled completely differently compared to a normal function.

File: web_src/js/features/repo-editor.js [hunk 13]
```diff
@@ -134,47 +134,49 @@ export async function initRepoEditor() {
   const $editArea = $('.repository.editor textarea#edit_area');
   if (!$editArea.length) return;
 
-  const editor = await createCodeEditor($editArea[0], $editFilename[0], previewFileModes);
+  (async () => {
+    const editor = await createCodeEditor($editArea[0], $editFilename[0], previewFileModes);
```
Comment (delvh) [inline web_src/js/features/repo-editor.js:138]: I am almost tempted to say this async is completely useless.
There is only one thing I can think of that would justify it: `createCodeEditor` is an incredibly slow function, and you don't want to block the main thread while waiting for it.
That's the only case where this would make at least some sense.
Is that the case?

File: web_src/js/features/common-global.js (hunk 5 above)
Comment (wxiaoguang) [reply web_src/js/features/common-global.js:142]: Because the callee is `async` but here do not not want to `await`. Using `const _promise` makes it clear.

File: web_src/js/features/comp/SearchUserBox.js (hunk 6 above)
Comment (wxiaoguang) [reply web_src/js/features/comp/SearchUserBox.js:5]: see other `initCompXxx`

File: web_src/js/features/repo-commit.js (hunk 8 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-commit.js:10]: see other `initRepoXxx`

File: web_src/js/features/repo-diff.js (hunk 9 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-diff.js:83]: Please search in the diff of this PR:  `import {initDiffShowMore} from './features/diff.js';`

File: web_src/js/features/repo-editor.js (hunk 10 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-editor.js:27]: The last comma is a good standard for all modern languages.

File: web_src/js/features/comp/SearchUserBox.js (hunk 6 above)
Comment (delvh) [reply web_src/js/features/comp/SearchUserBox.js:5]: Yes, I've seen it. But that doesn't mean I'm less confused about it.
What exactly is this supposed to convey?

File: web_src/js/features/repo-diff.js (hunk 9 above)
Comment (delvh) [reply web_src/js/features/repo-diff.js:83]: But you deleted that file? Or at least moved the method, so why rename it then?

File: web_src/js/features/repo-projects.js (hunk 12 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-projects.js]: Updated. If you do not like `const _promise = someAsyncFunc()`, you can do 

```
  (async () => {
    await initRepoProjectSortable();
  })();
```

They are the same.

File: web_src/js/features/notification.js (hunk 7 above)
Comment (wxiaoguang) [reply web_src/js/features/notification.js:69]: Explained above.

File: web_src/js/features/common-global.js (hunk 5 above)
Comment (wxiaoguang) [reply web_src/js/features/common-global.js:142]: Another way is:

```
(async()=>{
    await createDropzone(...);
})();
```

File: web_src/js/features/repo-issue-content.js (hunk 11 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-issue-content.js:107]: We must add these prefixes here, to make it clear which module does this function belong to.

There are global issue related features, and repo-level issue related features.

As long as we have meaningful prefixes, we can continue to group these functions together.

Otherwise, if you see a function `initIssueXxx`, how do you know whether you should call it for global init stage, or for a repo page init stage?

File: web_src/js/features/repo-editor.js (hunk 13 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-editor.js:138]: This is a correct case for `async/await`. Without `async () =>`, the `await` can not be used inside.

We must wait for editor to be created (asynchronously), it may be loaded via network request.

```
// equals to
(()=> {
    Promise.resolve(loadRemoteJs).then((editor) => {
         // now we get a usable editor object.
         editor.init(...)
         editor.setValue(...)
    });
})();
```

File: web_src/js/features/repo-diff.js (hunk 9 above)
Comment (wxiaoguang) [reply web_src/js/features/repo-diff.js:83]: Of course, moved.

File: web_src/js/features/comp/SearchUserBox.js (hunk 6 above)
Comment (wxiaoguang) [reply web_src/js/features/comp/SearchUserBox.js:5]: We must add these prefixes here, to make it clear which module does this function belongs to.

There are global issue related features, and repo-level issue related features.

As long as we have meaningful prefixes, we can continue to group these functions together.

Otherwise, if you see a function initIssueXxx, how do you know whether you should call it for global init stage, or for a repo page init stage?

File: web_src/js/features/notification.js (hunk 3 above)
Comment (wxiaoguang) [reply web_src/js/features/notification.js]: Explained in other comments.

Comment (wxiaoguang) [reply web_src/js/features/notification.js]: Because the callee is async but here do not not want to await. Using `const _promise` makes it clear.

Another way is:

```
(async()=>{
    await createDropzone(...);
})();
```

File: web_src/js/features/common-global.js (hunk 1 above)
Comment (lunny) [reply web_src/js/features/common-global.js:138]: The original code, `createDropzone` will run until all `initXXX` functions finished. So I guess module `dropzone` may has some dependencies. But I pull this PR and tested it, it works.

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
@silverwind need your help.
@go-gitea/maintainers any more advice?

**Review by techknowlogick — APPROVED** [review]

**Comment by wxiaoguang** [comment]
trigger bot
